product-feedback endpoint

npretto opened the pull request:
**Note: I'm not a BE dev so any feedback is appreciated** 🙏 

Part of  [[Epic] Initial homepage experience for embedding admins](https://github.com/metabase/metabase/issues/40005)
[Eng plan](https://www.notion.so/metabase/Initial-homepage-experience-for-embedding-admins-was-Meet-embedders-at-the-door-full-version-63fa3a8e07634c9faf67d5bdd5a2df08?pvs=4#dbde642c901749cfa439b3989f299ea9)

This PR introduces a new endpoint, `/api/utils/product-feedback` which acts as a proxy towards harbor master to send feedback from within Metabase.

The endpoint will be used initially from the [FeedbackModal for the Embedding Homepage](https://github.com/metabase/metabase/pull/40746).

By default, in dev mode it will use the value from an env as base domain, *even if empty*, this is by choice so that we don't send fake data to the real endpoint while testing the feature.

npretto committed:
cleanup and fixes

npretto committed:
mvp of the proxy endpoint for the product feedback

npretto commented on test/metabase/util/product_feedback_test.clj:
```diff
@@ -0,0 +1,34 @@
+                :error_details     "network issues",
+                :feedback_endpoint "http://test.example.org/api/v1/crm/product-feedback"}
+               (#'product-feedback/send-product-feedback "some feedback" "some-source" "some-email@example.org")))))
+    (testing "Should forward the data to the cloud endpoint"
```
With this test I wanted to make sure we're sending the correct data.
I tried to look for similar tests but I couldn't find any, any feedback is welcome.
Basically I wanted to check if I called http/post with the correct parameter, my solution was to throw an error in the handler if it received the wrong data, so that it returned `{"status":"success"}` only with the correct data

npretto commented on src/metabase/util/product_feedback.clj:
```diff
@@ -0,0 +1,52 @@
+(ns metabase.util.product-feedback
```
I based this off `fetch-token-and-parse-body`, I'm not entirely sure what some things I copied do so any feedback is welcome

filipesilva reviewed (commented):
If it's just forwarding couldn't the frontend just be calling store? Honest question, idk if that's a bad pattern in core.

npretto commented:
> If it's just forwarding couldn't the frontend just be calling store? Honest question, idk if that's a bad pattern in core.

That was one solution on the table, but [when we discussed](https://metaboat.slack.com/archives/C063Q3F1HPF/p1711550919323269) how to not point to the prod endpoint when playing around with mb in dev mode, using an ENV from the BE seemed like the cleanest solution.

piranha commented:
> If it's just forwarding couldn't the frontend just be calling store? Honest question, idk if that's a bad pattern in core.

In service world (when you run something and not the others) it usually makes lots of sense to make frontend do everything possible through backend, since it's much easier to control and understand. It's less important in a product like Metabase, but even that reason (dev vs prod) makes quite a bit of sense.

piranha commented on src/metabase/api/util.clj:
```diff
@@ -53,4 +54,13 @@
+  [:as {{:keys [comments source email]} :body}]
+  {comments ms/NonBlankString
+   source ms/NonBlankString
+   email [:maybe :string]}
```
`[:maybe ms/NonBlankString]` maybe?

npretto replied:
Do we have a strong preference for null over empty string? Is it because of how they're handled in the db?

npretto committed:
email :string -> ms/NonBlankString

dpsutton commented on src/metabase/api/util.clj:
```diff
@@ -53,4 +54,13 @@
+  {comments ms/NonBlankString
+   source ms/NonBlankString
+   email [:maybe ms/NonBlankString]}
+  (send-product-feedback comments source email))
```
A lot of times what we'll do is something like

```clojure
(future (send-product-feedback comments source email))
api/generic-204-no-content
```

This makes it clear that the handler won't block, just does it's little side effect forwarding, and immediately returns. I don't think there's any info we need to give to someone here except "thanks".

dpsutton commented:
I think this could be a lot simpler. My thinking:
- just put the function that makes a request in the api namespace. I don't think we need a new namespace to hold the one function.
- have the endpoint always return `api/generic-204-no-content`. I don't think we really want to forward errors to the user. We just bulk collect stuff and move on
- I'm interested to see what kind of error messages sending a null comment or a blank comment is. We might want custom checking instead of the automated type checking. Or maybe changing it to just a "string" and then throw a nice error message if it's blank.
- The tests don't really do much, or assert a behavior that I'm not sure that we actually want.
- The behavior that we care about really is that we forward the request with three bits of information. We should probably assert that, that we have some decent error messages, and if there's network problems we don't block or annoy the user.

dpsutton commented:
I think a simpler version is this patch based on master:

```diff
diff --git a/src/metabase/api/util.clj b/src/metabase/api/util.clj
index 380cda08ee..7921e97ae0 100644
--- a/src/metabase/api/util.clj
+++ b/src/metabase/api/util.clj
@@ -2,6 +2,8 @@
   "Random utilty endpoints for things that don't belong anywhere else in particular, e.g. endpoints for certain admin
   page tasks."
   (:require
+   [cheshire.core :as json]
+   [clj-http.client :as http]
    [compojure.core :refer [GET POST]]
    [crypto.random :as crypto-random]
    [metabase.analytics.prometheus :as prometheus]
@@ -9,7 +11,10 @@
    [metabase.api.common :as api]
    [metabase.api.common.validation :as validation]
    [metabase.logger :as logger]
+   [metabase.models.setting :refer [defsetting]]
    [metabase.troubleshooting :as troubleshooting]
+   [metabase.util.log :as log]
+   [metabase.util.malli :as mu]
    [metabase.util.malli.schema :as ms]
    [ring.util.response :as response]))
 
@@ -38,6 +43,33 @@
   []
   {:token (crypto-random/hex 32)})
 
+(defsetting product-feedback-url
+  "URL for where to send product feedback"
+  :type :string
+  :default "https://store-api.metabase.com/api/v1/crm/product-feedback"
+  :visibility :internal)
+
+(mu/defn send-feedback!
+  [comments :- ms/NonBlankString
+   source :- ms/NonBlankString
+   email :- ms/NonBlankString]
+  (try (http/post (product-feedback-url)
+                  {:content-type :json
+                   :body         (json/generate-string {:comments comments
+                                                        :source   source
+                                                        :email    email})})
+       (catch Exception e
+         (log/warn e)
+         (throw e))))
+
+(api/defendpoint POST "/product-feedback"
+  [:as {{:keys [comments source email]} :body}]
+  {comments ms/NonBlankString
+   source ms/NonBlankString
+   email ms/NonBlankString}
+  (future (send-feedback! comments source email))
+  api/generic-204-no-content)
+
 (api/defendpoint GET "/bug_report_details"
   "Returns version and system information relevant to filing a bug report against Metabase."
   []
diff --git a/test/metabase/api/util_test.clj b/test/metabase/api/util_test.clj
index 5c8c583240..a917182098 100644
--- a/test/metabase/api/util_test.clj
+++ b/test/metabase/api/util_test.clj
@@ -1,7 +1,9 @@
 (ns metabase.api.util-test
   "Tests for /api/util"
   (:require
+   [clojure.string :as str]
    [clojure.test :refer :all]
+   [metabase.api.util :as api.util]
    [metabase.test :as mt]
    [metabase.util.log :as log]))
 
@@ -57,3 +59,24 @@
              (mt/user-http-request :rasta :get 403 "util/diagnostic_info/connection_pool_info"))))
     (testing "Call successful for superusers"
       (is (map? (mt/user-http-request :crowberto :get 200 "util/diagnostic_info/connection_pool_info"))))))
+
+(deftest product-feedback-tests
+  (testing "requires non-blank comments, source, and email"
+    (doseq [missing [:comments :source :email]]
+      (let [payload  {:comments "foo" :source "foo" :email "foo"}
+            response (mt/user-http-request :crowberto :post 400 "util/product-feedback" (dissoc payload missing))]
+        (testing (str "without " missing)
+          (is (= {:errors          {missing "value must be a non-blank string."},
+                  :specific-errors {missing ["should be a string, received: nil" "non-blank string, received: nil"]}}
+                 response))))))
+  (testing "fires the proxy in background"
+    (let [sent? (promise)]
+      (with-redefs [api.util/send-feedback! (fn [comments source email]
+                                              (doseq [prop [comments source email]]
+                                                (is (not (str/blank? prop)) "got a blank property to send-feedback!"))
+                                              (deliver sent? true))]
+        (mt/user-http-request :crowberto :post 204 "util/product-feedback"
+                              {:comments "I like Metabase"
+                               :email    "happy_user@test.com"
+                               :source   "Analytics Inc"})
+        (is (true? (deref sent? 2000 ::timedout)))))))
```

I think this is far simpler, a few better tests, still easy to override the URL, and doesn't add extra files. Also, I checked what kind of errors we get from the ms/nonblankstring and they seem ok. I have tests about their shape and have an example here:

```shell
❯ http post 'localhost:3000/api/util/product-feedback' x-api-key:$API_KEY -pb
{
    "errors": {
        "comments": "value must be a non-blank string."
    },
    "specific-errors": {
        "comments": [
            "should be a string, received: nil",
            "non-blank string, received: nil"
        ]
    }
}
```

camsaul commented on src/metabase/util/product_feedback.clj:
```diff
@@ -0,0 +1,52 @@
+(defn send-product-feedback
+  "Send product feedback from the app"
+  [comments source email]
+  (let [fut    (future (send-product-feedback* comments source email))
+        result (deref fut send-feedback-timeout-ms ::timed-out)]
+    (if (not= result ::timed-out)
```
We have a util function that does this, look for usages of `u/with-timeout` (it's defined in `metabase.util.jvm/with-timeout`)

camsaul commented on src/metabase/util/product_feedback.clj:
```diff
@@ -0,0 +1,52 @@
+
+(set! *warn-on-reflection* true)
+
+ (def product-feeback-base-url
```
You should use `defsetting` for this, then you get the `MB_PRODUCT_...` stuff for free and you can use all of our tooling around mocking settings. Something like

```clj
(defsetting product-feeback-base-url
  "Base url for the product feedback, in DEV it uses a value from an ENV, even if empty, to not send real data to the production endpoint"
  :visibility :internal
  :default ...)
```

camsaul commented on src/metabase/util/product_feedback.clj:
```diff
@@ -0,0 +1,52 @@
+       :error_details     (.getMessage e)
+       :feedback_endpoint product-feeback-url})))
+
+(defn send-product-feedback
```
```suggestion
(defn send-product-feedback!
```

Since it has side effects. It would be different to to call the function twice instead of once, not a pure function

WiNloSt commented:
I've reviewed it, but didn't have much opinion on the BE code. It's ok as long as it works for me.

npretto committed:
revert my solution

npretto committed:
Dan diff file

npretto committed:
makes fields other than source optional

npretto committed:
don't use default value in dev, to not spam the prod endpoint

npretto commented:
Thank you @dpsutton and @camsaul !
I reverted my changes and started again from Dan's patch (It's on its own commit so you can see my diffs more easily).

I changed a couple of things:
- both email and comments are now optional, it was my mistake to mark comments at required 
- I set a `setter` on the setting to make it not point to prod on dev by default

The second point is because I don't want us to send fake data points to the real endpoint accidentally, do you have better solutions to that?
Not sure if the setting is still worth it as I still had to write the setter which is pretty much what could be a variable definition

npretto commented on src/metabase/api/util.clj:
```diff
@@ -47,12 +47,14 @@
   :default "https://store-api.metabase.com/api/v1/crm/product-feedback"
-  :visibility :internal)
+  :visibility :internal
+  :export? false)
```
Linter was complaining about this, I set it to false as I don't think this will actually be changed by any real user

dpsutton replied:
Can you tell me your thinking behind this custom getter?

npretto replied:
I don't want people to start metabase with `yarn dev` and accidentally send feedback to the real endpoint when playing around with it.
I know we can still customize it with the env, but without the custom getter it would still default to the prod endpoint

npretto replied:
Not sure if the setting is still worth it as the setter is basically what could go in a (def var ...)

dpsutton replied:
ah i see!. So we don't want to invite anyone to change this, and we want to make sure that at dev time it is intentional. I think this might be better then:

```clojure
(defn- product-feedback-url
  "Product feedback url. When not prod, reads `MB_PRODUCT_FEEDBACK_URL` from the environment to prevent development
  feedback from hitting the endpoint."
  []
  (if config/is-prod?
    "https://store-api.metabase.com/api/v1/crm/product-feedback"
    (env :mb-product-feedback-url)))

(mu/defn send-feedback!
  "Sends the feedback to the api endpoint"
  [comments :- [:maybe ms/NonBlankString]
   source :- ms/NonBlankString
   email :- [:maybe ms/NonBlankString]]
  (try (http/post (product-feedback-url)
                  {:content-type :json
                   :body         (json/generate-string {:comments comments
                                                        :source   source
                                                        :email    email})})
       (catch Exception e
         (log/warn e)
         (throw e))))
```

this changes the `defsetting` to be a function. As a setting, it seems to invite people to change it, and what we really want to do is always use one during production and require a dev to set a value at dev time.

Does this seem right?

npretto replied:
That's perfect thanks 🙏 , pushed it in https://github.com/metabase/metabase/pull/40974/commits/2c27d07800b52df757fa137c8b442c3948bd8df5

npretto committed:
change from setting to function for product-feedback-url, as it's not meant to be changed by end users
